File: 剑指offer/Codes/14_枪打出头鸟_栈.cpp
using namespace std;

#include<iostream>
#include<vector>

class Solution {
public:
    /**
     * 
     * @param n int整型 n个人
     * @param a int整型vector ai代表第i个人的高度
     * @return long长整型
     */
    long long solve(int n, vector<int>& a) {
        int i = n-1;
        long long SUM = 0;
        while (i > 0){
            for (int j = i-1; j >= 0; j--){
                if (a[j] > a[i]){
                    SUM += j+1;
                    break;
                }
            }
            i --;
        }
        return SUM;
    }
};


int main(){
    int n = 5;
    Solution s = Solution();
    // int a[5] = {1, 2, 3, 4, 5};
    int a[5] = {5, 4, 3, 2, 1};
    vector<int> obj;
    for(int i=0; i < n; i++){
        obj.push_back(a[i]);
    }
    cout << s.solve(n, obj) << endl;
}

File: 剑指offer/Codes/14_枪打出头鸟_栈.py
#
# 
# @param n int整型 n个人
# @param a int整型一维数组 ai代表第i个人的高度
# @return long长整型
#
class Solution:
    def solve(self , n , a ):
        # write code here
        index = n - 1
        SUM = 0
        while index > 0:
            temp_index = index - 1
            while a[temp_index] < a[index]:
                # 如果已经到了第一个位置，还没有找到比自己高的元素，则自己是最高的
                if temp_index == 0:
                    SUM += 0
                    break    
                temp_index -= 1
                
            SUM += temp_index + 1 
            print("SUM", SUM, temp_index + 1)

            index -= 1
 
        return SUM
            
# print(Solution().solve(5,[1,2,3,4,5]))
# print(Solution().solve(5,[5,4,3,2,1]))